File: offload.f90
PROGRAM offload

USE omp_lib
IMPLICIT NONE

INTEGER, PARAMETER :: Niter = 100  !Number of loop iterations
INTEGER, PARAMETER :: SIZE = 64000 !Size of arrays

!variables for looping and timing
INTEGER :: i, tstart, tend, trate

DOUBLE PRECISION, DIMENSION(SIZE) :: in_array, out_array

   ! Initialize arrays
   DO i = 1,SIZE
      in_array(i) = float(i+1)
      out_array(i) = 0
   ENDDO

   !Start the timer
   call system_clock(count_rate=trate) !Find the time rate
   call system_clock(count=tstart)     !Start Timer
   
   DO i = 1,Niter
!Pragma for offloading work onto the target
!WORKSHOP: Offload some_work to the target using an appropriate pragma
!The array in_array will be copied from host->target at the start
!out_array will be copied from target->host at the end
      call some_work(in_array, out_array)
   ENDDO

   !Stop the clock
   call system_clock(count=tend)      ! Stop Timer


   !Print time
   WRITE(*,*) 'Time: ', real(tend-tstart)/real(trate), &
       ' s'

!This function will be compiled only for the target device (mic)

!WORKSHOP: Declare some_work for device execution instead of 
! host execution
CONTAINS
SUBROUTINE some_work(in, out)
!Computes the sin squared for each element of in
   INTEGER :: i
   DOUBLE PRECISION :: sinx
   DOUBLE PRECISION, DIMENSION(SIZE) :: in, out
!OpenMP pragma for parallel for loop
!$OMP PARALLEL DO
   DO i = 1, SIZE
      sinx = sin( in(i) )
      out(i) = sinx*sinx
   ENDDO
!$OMP END PARALLEL DO
END SUBROUTINE some_work
END PROGRAM offload

File: offload.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <math.h>

#define Niter 100  //Number of loop iterations
#define SIZE 64000 //Size of arrays

//WORKSHOP: Declare some_work so that it is compiled 
// for the target instead of the host
void some_work(float *in, float *out)
//Computes the sin squared for each element of in
{
   int i;
//OpenMP pragma for parallel for loop
//private(i) indicates each thread will have a private variable i
#pragma omp parallel for private(i)
   for (i = 0; i < SIZE; i++)
   {
      double sinx = sin( in[i] );
      out[i] = sinx*sinx;
   }
}

double get_time()
//Returns the current time in seconds
{
   struct timeval tp;
   gettimeofday(&tp, NULL);
   return (double)(tp.tv_sec) + (double)(tp.tv_usec)/1E6;
}


int main(int argc, char* argv[])
{
   int i;
   double tstart, tend;
   float *in_array, *out_array;

   //  Allocate arrays and check for errors
   in_array = (float*)malloc(SIZE*sizeof(float));
   out_array = (float*)malloc(SIZE*sizeof(float));
   if (!in_array || !out_array)
   {
      printf("*** Error - Allocation error\n");
      return 0;
   }

   // Initialize arrays
   for (i = 0; i < SIZE; i++)
   {
      in_array[i] = (float) (i + 1);
      out_array[i] = 0;
   }

   //Start the timer
   tstart = get_time();
   
   for(i = 0; i < Niter; i++)
   {
     //WORKSHOP: Offload some_work to the target using an appropriate pragma
     // in_array should be copied from host to target at beginning
     // out_array should be copied from target to host at the end
       some_work(in_array, out_array);
   }
   //Stop the clock
   tend = get_time();

   // Print time
   printf("Time:  %.3lf seconds\n", (double) tend - tstart);
   free(in_array);
   free(out_array);

   return 0;
}
